src/weapons.py

```diff
@@ -18,10 +22,52 @@ class DamageType(Enum):
     thunder = "thunder"
 
 
+@dataclass
+class RollAffects:
+    immunities: List[DamageType]
+    resistances: List[DamageType]
+    vulnerabilities: List[DamageType]
+
+
+class CompoundRollResult:
+    def __init__(self) -> None:
+        self._rolls: Dict[DamageType, List[RollResult]] = defaultdict(lambda _: [])
+
+    def add_roll(self, damage_type: DamageType, roll: RollResult) -> None:
+        self._rolls[damage_type].append(roll)
+
+    def calculate_roll(self, roll_affects: RollAffects) -> int | float:
+        uncalculated_rolls = self._rolls.copy()
+        total_damage: int | float = 0
+        for immunity in roll_affects.immunities:
+            uncalculated_rolls.pop(immunity, None)
+        for resistance in roll_affects.resistances:
+            for roll in uncalculated_rolls.pop(resistance, []):
+                uncalculated_rolls += roll.result / 2
+        for vulnerability in roll_affects.vulnerabilities:
+            for roll in uncalculated_rolls.pop(vulnerability, []):
+                uncalculated_rolls += roll.result * 2
+        # Calculate the remaining damage types
+        for damage_type, rolls in uncalculated_rolls:
+            for roll in rolls:
+                uncalculated_rolls += roll.result
+        return total_damage
+
+
 class Weapon:
-    def __init__(self, name: str, damage_die: Die):
+    def __init__(self, name: str, damage_dice: List[Die], damage_type: DamageType):
```

Comment:
I have yet to see a weapon with several die types for damage, but if you see us adding such custom weapons to the game, I'm down

Reply:
That was defined in the issue. For example, a dart covered in venom might have both poison and piercing damages

Reply:
a list of Die doesn't give you that information. What am I missing?
Maybe it should require CompoundDamageRollResult and it can replace the variable on [line 67](https://github.com/theUltimateZoltan/dnd_utils/blob/e267147751a3d8c42f7744d8aae4438225afc843/src/weapons.py#L67).